UefiPayloadPkg/Library/BuildFdtLib/X86_BuildFdtLib.c

```diff
@@ -0,0 +1,980 @@
+    ASSERT_EFI_ERROR (Status);
+
+    RegData[0] = CpuToFdt64 (GraphicsInfo->FrameBufferBase);
+    RegData[1] = CpuToFdt64 (GraphicsInfo->FrameBufferSize);
+    Status     = FdtSetProp (Fdt, TempNode, "reg", &RegData, sizeof (RegData));
+    ASSERT_EFI_ERROR (Status);
+
+    Status = FdtSetProp (Fdt, TempNode, "compatible", "simple-framebuffer", (UINT32)(AsciiStrLen ("simple-framebuffer")+1));
+    ASSERT_EFI_ERROR (Status);
+  } else {
+    Status   = AsciiSPrint (TempStr, sizeof (TempStr), "framebuffer@%lX", 0xB0000000);
+    TempNode = FdtAddSubnode (Fdt, 0, TempStr);
+    ASSERT (TempNode > 0);
+
+    Status = FdtSetProp (Fdt, TempNode, "display", "&gma", (UINT32)(AsciiStrLen ("&gma")+1));
+    ASSERT_EFI_ERROR (Status);
+
+    Status = FdtSetProp (Fdt, TempNode, "format", "a8r8g8b8", (UINT32)(AsciiStrLen ("a8r8g8b8")+1));
+    ASSERT_EFI_ERROR (Status);
+
+    Data32 = CpuToFdt32 (1024);
+    Status = FdtSetProp (Fdt, TempNode, "height", &Data32, sizeof (UINT32));
+    ASSERT_EFI_ERROR (Status);
+
+    Data32 = CpuToFdt32 (1280);
+    Status = FdtSetProp (Fdt, TempNode, "width", &Data32, sizeof (UINT32));
+    ASSERT_EFI_ERROR (Status);
+
+    RegData[0] = CpuToFdt64 (0xB0000000);
+    RegData[1] = CpuToFdt64 (0x500000);
+    Status     = FdtSetProp (Fdt, TempNode, "reg", &RegData, sizeof (RegData));
+    ASSERT_EFI_ERROR (Status);
+
+    Status = FdtSetProp (Fdt, TempNode, "compatible", "simple-framebuffer", (UINT32)(AsciiStrLen ("simple-framebuffer")+1));
+    ASSERT_EFI_ERROR (Status);
+  }
+
+  return Status;
+}
+
+/**
+  It will build FDT for UPL required data.
+  @param[in] FdtBase         Address of the Fdt data.
+  @retval EFI_SUCCESS        If it completed successfully.
+  @retval Others             If it failed to build required FDT.
+**/
+EFI_STATUS
+BuildFdtForUplRequired (
+  IN     VOID  *FdtBase
+  )
+{
+  EFI_STATUS              Status;
+  VOID                    *Fdt;
+  VOID                    *Fit;
+  INT32                   ParentNode;
+  INT32                   CustomNode;
+  INT32                   UPLParaNode;
+  INT32                   UPLImageNode;
+  EFI_HOB_CPU             *CpuHob;
+  UINT64                  Data64;
+  UINT32                  Data32;
+  UINT16                  Data16;
+  VOID                    *HobPtr;
+  EFI_BOOT_MODE           BootMode;
+  CHAR8                   TempStr[32];
+  UINT8                   *GuidHob;
+  UNIVERSAL_PAYLOAD_BASE  *PayloadBase;
+
+  Fdt = FdtBase;
+  Fit = NULL;
+
+  //
+  // Create Hob list FDT node.
+  //
+  ParentNode = FdtAddSubnode (Fdt, 0, "options");
+  ASSERT (ParentNode > 0);
+
+  UPLParaNode = FdtAddSubnode (Fdt, ParentNode, "upl-params");
+  ASSERT (UPLParaNode > 0);
+
+  //
+  // Create CPU info FDT node
+  //
+  CpuHob = GetFirstHob (EFI_HOB_TYPE_CPU);
+  ASSERT (CpuHob != NULL);
+
+  Data16 = CpuToFdt16 (0);
+  Status = FdtSetProp (Fdt, UPLParaNode, "pci-enum-done", &Data16, sizeof (Data16));

```

Comment:
pci-enum-done is boolean type data, we do not need to assign value to it.
please try with "len=0" and "value input as NULL", see if we could create empty property.

Reply:
resolved.